packages/whenable/src/whenable.js

```diff
@@ -0,0 +1,111 @@
+import { makeTagged } from '@endo/pass-style';
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ */
+export const prepareWhenableKits = zone => {
+  /** WeakMap<object, any> */
+  const whenableV0ToEphemeral = new WeakMap();
+
+  /**
+   * Get the current incarnation's promise kit associated with a whenableV0.
+   *
+   * @param {import('./types.js').WhenablePayload['whenableV0']} whenableV0
+   * @returns {import('@endo/promise-kit').PromiseKit<any>}
+   */
+  const findCurrentKit = whenableV0 => {
+    let pk = whenableV0ToEphemeral.get(whenableV0);
+    if (pk) {
+      return pk;
+    }
+
+    pk = makePromiseKit();
+    pk.promise.catch(() => {}); // silence unhandled rejection
+    whenableV0ToEphemeral.set(whenableV0, pk);
+    return pk;
+  };
+
+  /**
+   * @param {'resolve' | 'reject'} kind
+   * @param {import('./types.js').WhenablePayload['whenableV0']} whenableV0
+   * @param {unknown} value
+   */
+  const settle = (kind, whenableV0, value) => {
+    const kit = findCurrentKit(whenableV0);
+    const cb = kit[kind];
+    if (!cb) {
+      return;
+    }
+    whenableV0ToEphemeral.set(whenableV0, harden({ promise: kit.promise }));
+    cb(value);
+  };
+
+  const makeWhenableInternalsKit = zone.exoClassKit(
+    'WhenableInternalsKit',
+    {
+      whenableV0: M.interface('WhenableV0', {
+        shorten: M.call().returns(M.promise()),
+      }),
+      settler: M.interface('Settler', {
+        resolve: M.call().optional(M.any()).returns(),
+        reject: M.call().optional(M.any()).returns(),
+      }),
+    },
+    () => ({}),
+    {
+      whenableV0: {
+        /**
+         * @returns {Promise<any>}
+         */
+        shorten() {
+          return findCurrentKit(this.facets.whenableV0).promise;
+        },
+      },
+      settler: {
+        /**
+         * @param {any} [value]
+         */
+        resolve(value) {
+          const { whenableV0 } = this.facets;
+          settle('resolve', whenableV0, value);
+        },
+        /**
+         * @param {any} [reason]
+         */
+        reject(reason) {
+          const { whenableV0 } = this.facets;
+          settle('reject', whenableV0, reason);
+        },
+      },
+    },
+  );
+
+  /**
+   * @template T
+   * @returns {import('./types.js').WhenableKit<T>}
+   */
+  const makeWhenableKit = () => {
+    const { settler, whenableV0 } = makeWhenableInternalsKit();
+    const whenable = makeTagged('Whenable', harden({ whenableV0 }));
+    return harden({ settler, whenable });
+  };
+
+  /**
+   * @template T
+   * @returns {import('./types.js').WhenablePromiseKit<T>}
+   */
+  const makeWhenablePromiseKit = () => {
+    const { settler, whenableV0 } = makeWhenableInternalsKit();
+    const whenable = makeTagged('Whenable', harden({ whenableV0 }));
+
+    /** @type {{ promise: Promise<T> }} */
+    const { promise } = findCurrentKit(whenableV0);
+    return harden({ settler, whenable, promise });
+  };
+  return { makeWhenableKit, makeWhenablePromiseKit };
```

Comment:
Rather than duplicating-with-variation `makeWhenableKit` as a `makeWhenablePromiseKit` for module-internal use, perhaps it would make more sense to expose `findCurrentKit` and refactor it to use the "settler" facet.
```suggestion
  return { makeWhenableKit, providePromiseForSettler: findCurrentKit };
```
and then in packages/whenable/src/when.js:
```diff
-    const { settler, promise } = makeWhenablePromiseKit();
+    const { settler } = makeWhenableKit();
+    const promise = providePromiseForSettler(settler);
```

Reply:
Nice!

Reply:
By "expose `findCurrentKit`", do you mean to expose it as public API of the package? Please no.

Reply:
> By "expose `findCurrentKit`", do you mean to expose it as public API of the package? Please no.

No, just internally from whenable.js for use by when.js.